src/usdb_syncer/gui/mw.py

```diff
@@ -269,6 +272,83 @@ def _generate_song_pdf(self) -> None:
         if path:
             generate_song_pdf(self.table.all_local_songs(), path)
 
+    def _import_usdb_ids_from_files(self) -> None:
+        logger = logging.getLogger()
+        file_list = QFileDialog.getOpenFileNames(
+            self,
+            caption="Select one or more files to import USDB IDs from",
+            dir=os.getcwd(),
+            filter="JSON, USDB IDs, Weblinks (*.json *.usdb_ids *.url *.webloc *.desktop)",
+        )[0]
+        file_parsers = [USDBIDFileParser(path) for path in file_list]
+        has_error = False
+        for parser in file_parsers:
+            if parser.errors:
+                logger.error(f"importing file {parser.filepath}: {parser.errors}")
+                has_error = True
+        # stop import if encounter errors
+        if has_error:
+            return
+        song_ids = [id for parser in file_parsers for id in parser.ids]
+        unique_song_ids = list(set(song_ids))
+        unique_song_ids.sort(key=lambda song_id: song_id.value)
+        logger.info(
+            f"read {len(file_list)} file(s), "
+            f"found {len(unique_song_ids)} "
+            f"USDB IDs: {[str(id) for id in unique_song_ids]}"
+        )
+        songs = [
+            DownloadInfo.from_song_data(song)
+            for song_id in unique_song_ids
+            if (song := self.table.get_data(song_id))
+        ]
+        if unavailable_song_ids := [
+            song_id for song_id in unique_song_ids if not self.table.get_data(song_id)
+        ]:
+            logger.warning(
+                f"{len(unavailable_song_ids)}/{len(unique_song_ids)} "
+                "imported USDB IDs are not available: "
+                f"{[str(song_id) for song_id in unavailable_song_ids]}"
+            )
+            if len(unavailable_song_ids) < len(unique_song_ids):
+                # select available songs to prepare Download
+                available_song_ids = [
+                    song_id
+                    for song_id in unique_song_ids
+                    if song_id not in unavailable_song_ids
+                ]
+                logger.info(
+                    f"available {len(available_song_ids)}/{len(unique_song_ids)} "
+                    "imported USDB IDs are selected for Download: "
+                    f"{[str(song_id) for song_id in available_song_ids]}"
+                )
+                self.table.set_selection_to_song_ids(available_song_ids)
+        else:
+            download_songs(
```

Comment:
My original idea was to simplify download of many specific files. I am using this features for the past weeks already and I like that the download just starts when there are no errors or warnings. Just adding to download list would need an additional interaction and the user might also accidentially click in the list of all songs, thereby reselect and the imported selection is gone.
I think you have a specific concept in mind, so if you still prefer I will change the behavior, but I wanted point out why I wanted to go for the direct download.

Reply:
I understand how it's easier for you, but other users will not be as familiar with this feature and may appreciate the additional intermediate step. Also, you currently only import your own files and know exactly what's inside them. But if users share id files in the future, it may be even more helpful if they can decide to only download a part of the imported ids.

> the user might also accidentially click in the list of all songs, thereby reselect and the imported selection is gone.

That's true, selecting has some problems. However, I'm currently working on a PR that adds a separate download batch table. So when I'm suggesting to _select songs_, in my mind that's actually already _adding the songs to the batch table._